### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by TSholokhova**
Trie implementation for Dictionary class.
Add new word O(n).
Search word O(n).

Cache memory implemented in Trie for better performance in word_search function. So, is_prefix and in_search functions have asymptotic O(1) when they called from word_search.

### CODE REVIEW

File: TSholokhova/assignment3/code.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+
+    def __init__(self, table):
+        self.grid = table
+        self.rows = 0
+        self.cols = 0
+        if table is not None:
+            self.rows = len(table)
+            self.cols = len(table[0])
+
+    def get_adjacent_cells(self, x0, y0):
```
Comment (ferreron) [inline TSholokhova/assignment3/code.py]: This is a good candidate for a generator: https://wiki.python.org/moin/Generators

File: TSholokhova/assignment3/code.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+            {END!}}
+          {END!}}
+        {T :
+          {END!}}}}
+    """
+    def __init__(self, words=None, use_cache=False):
+        """
+        :param words: initial words, list of strings.
+        :param use_cache: use "cache memory" to avoid recomputation of the same values.
+        When use_cache is True methods is_word and is_prefix should not receive a string.
```
Comment (ferreron) [inline TSholokhova/assignment3/code.py]: Can you explain in your comments a bit more why you use a cache, what the advantage is, etc.?

File: TSholokhova/assignment3/code.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+            Results of these methods depends only on cache_root.
+            It is necessary use navigation cash methods: init, add, del.
+        """
+        self.root = dict()
+        self.use_cache = use_cache
+        if self.use_cache:
+            self.init_cache()
+        if not words:
```
[the remark this answers is not in the document]
Comment (TSholokhova) [reply TSholokhova/assignment3/code.py]: Done

File: TSholokhova/assignment3/code.py (hunk 1 above)
Comment (TSholokhova) [reply TSholokhova/assignment3/code.py]: Done

File: TSholokhova/assignment3/code.py (hunk 2 above)
Comment (TSholokhova) [reply TSholokhova/assignment3/code.py]: Done

**Pull request merged**